src/Microsoft.AspNet.Server.Kestrel/Http/SocketOutput.cs

```diff
@@ -80,23 +78,25 @@ public SocketOutput(KestrelThread thread, UvStreamHandle socket, long connection
                 {
                     _nextWriteContext.SocketDisconnect = true;
                 }
+
                 // Complete the write task immediately if all previous write tasks have been completed,
                 // the buffers haven't grown too large, and the last write to the socket succeeded.
-                triggerCallbackNow = _lastWriteError == null &&
-                                     _callbacksPending.Count == 0 &&
-                                     _numBytesPreCompleted + buffer.Count <= _maxBytesPreCompleted;
-                if (triggerCallbackNow)
+                if (_lastWriteError == null &&
+                    _tasksPending.Count == 0 &&
+                    _numBytesPreCompleted + buffer.Count <= _maxBytesPreCompleted)
                 {
                     _numBytesPreCompleted += buffer.Count;
                 }
+                else if (immediate)
+                {
+                    // immediate write, which is not eligable for instant completion above
+                    tcs = new TaskCompletionSource<object>(buffer.Count);
+                    _tasksPending.Enqueue(tcs);
+                }
                 else
                 {
-                    _callbacksPending.Enqueue(new CallbackContext
-                    {
-                        Callback = callback,
-                        State = state,
-                        BytesToWrite = buffer.Count
-                    });
+                    // immediate==false calls always return complete tasks, because there is guaranteed
+                    // to be a subsequent immediate==true call which will go down one of the previous code-paths
```

Comment:
I've discovered my profiling (in t'Cloud) is slightly "time of day" effected; so not great at testing smaller changes :(

And larger changes aren't great for PRs; so trying to break the larger one into chunks. Will paste some memory profiles while waiting for Travis to work...